[CONTP-388] use entityid struct in tagger component

adel121 opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

This PR changes the signature of the `Tag` method of the tagger interface to receive the entity id as a struct instead of a string.

### Motivation

TL:DR; Avoid string manipulation overhead.

Full context:
Entity ID is composed of a prefix and an id and is displayed as `{prefix}://{id}`.
Callers of the tagger usually have the prefix and the id parts separated, and they form the entityID as a string by concatenating the prefix and id along with the separator: `fmt.Sprintf("%s://%s", prefix, id)`. 
This entityID is passed as a string to the tagger Tag method to get the entity tags. 
However, the tagger needs to separate the prefix and the id, so it will need to parse the string, adding an overhead.

To eliminate this overhead, with the change introduced in this PR, the entityID will now be passed as a struct containing 2 fields, one for the prefix and one for the id, eliminating the need to perform string manipulations (concatenations and splits). 

### Describe how to test/QA your changes

Unit tests and E2E tests should cover most the QA.

However, we still need to verify that python integrations can still query the tagger normally.

Here is a sample QA:

Deploy the agent with helm.

Do the following steps to deploy a custom python check onto the agent:

1. Check the running pods:
```
kubectl get pods
NAME                                         READY   STATUS    RESTARTS      AGE
datadog-agent-cluster-agent-cfc87c6c-d26pg   1/1     Running   0             107m
datadog-agent-hsstl                          3/3     Running   4 (73m ago)   107m
```

2. Check the content of the tagger, and pick any tag entity id for testing:
```
kubectl exec datadog-agent-hsstl -- agent tagger-list
....
=== Entity kubernetes_pod_uid://b97232df7635c1e88c7bd9309b8e78ae ===
== Source workloadmeta-kubernetes_pod =
=Tags: [kube_namespace:kube-system kube_priority_class:system-node-critical kube_qos:Burstable pod_name:kube-apiserver-kind-control-plane pod_phase:running]
===
....
```

3. Exec into the agent container: `kubectl exec -it datadog-agent-hsstl -c agent -- bin/bash`
4. Add a file called `custom_checkvalue.yaml` under `etc/datadog-agent/conf.d` with the following content:
```
init_config:
instances:
  [{}]
```
5. Add the check code in a file called `custom_checkvalue.py` under `etc/datadog-agent/checks.d` with the following content:

```
from datadog_checks.base.utils.tagging import tagger
from checks import AgentCheck
class HelloCheck(AgentCheck):
  def check(self, instance):
    tags = tagger.tag('kubernetes_pod_uid://b97232df7635c1e88c7bd9309b8e78ae', tagger.HIGH)
    print("obtained tags: ", tags)
    self.gauge('hello.world', 1, tags)
```

ATTENTION: you should use the entity id that you go from the tagger-list command in step 2.

6. Restart the agent by running: `agent stop && agent start`

Wait for the agent to start, and run the custom check as follows:

```
 kubectl exec datadog-agent-hsstl -- agent check custom_checkvalue > output.txt

....
=== Series ===
[
  {
    "metric": "hello.world",
    "points": [
      [
        1728996360,
        1
      ]
    ],
    "tags": [
      "kube_namespace:kube-system",
      "kube_priority_class:system-node-critical",
      "kube_qos:Burstable",
      "pod_name:kube-apiserver-kind-control-plane",
      "pod_phase:running"
    ],
    "host": "kind-control-plane-adel-test-rtloader",
    "type": "gauge",
    "interval": 0,
    "source_type_name": "System"
  }
]
....
```

Note: you migh not get the tags on each run due to lag of initialisation of workloadmeta and tagger. However, as long as you are getting the tags at least once, you can consider the QA as done. You can also check the Datadog UI to validate that you can see the metric and aggregate it by tags:

![image](https://github.com/user-attachments/assets/094bd851-7d8e-484f-a991-c34a29e5726e)


### Possible Drawbacks / Trade-offs
Rtloader exposes the tagger tag method to the python integrations running in the datadog agent. 

Changing the signature of the Tag method breaks the rtloader interface, and thus breaks the calls to the tagger from the python integrations. 

To preserve backward compatibility in regards for rtloader, the old signature is kept in the codebase, but now with the name `LegacyTag`, and the rtloader now redirects python calls to the tagger Tag method to the `LegacyTag` function.

The `LegacyTag` function receives the entityID as a string, parses the string to split it to a prefix and and an id, and redirects the call to the `Tag` function.

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

adel121 committed:
use entityID struct in tagger Tag method

adel121 committed:
update tagger to use entityid struct to return tags

adel121 committed:
update calls to the tagger

adel121 commented on comp/core/tagger/global.go:
```diff
@@ -44,8 +44,25 @@ func GetEntity(entityID types.EntityID) (*types.Entity, error) {
+// If possible, avoid using this function, and use the Tag interface function instead.
+// This function exists in order not to break backward compatibility with rtloader and python
+// integrations using the tagger
+func LegacyTag(entity string, cardinality types.TagCardinality) ([]string, error) {
```
I see your point, but the thing here is that we need to push developers away from using this method because it includes an extra step of string parsing, which we can avoid by calling `Tag` method of the tagger. 

This is why I think it is good to include something in the function name indicating that calling it should be avoided except in very specific cases. 

How about: 

`LegacyTagByString` ?

L3n41c commented on comp/core/tagger/global.go:
```diff
@@ -44,8 +44,25 @@ func GetEntity(entityID types.EntityID) (*types.Entity, error) {
+// integrations using the tagger
+func LegacyTag(entity string, cardinality types.TagCardinality) ([]string, error) {
+	if globalTagger == nil {
+		return nil, fmt.Errorf("a global tagger must be set before calling Tag")
```
`fmt.Errorf` is a function aiming at formatting a string.
When there’s nothing to format, it’s much more efficient to use a function that just wraps the already formatted string into an `error`:

```suggestion
		return nil, errors.New("a global tagger must be set before calling Tag")
```

adel121 replied:
Indeed you are right. 

fmt.Errorf is already used in this file in many place, where it can be replaced by `errors.New`. 

I updated all of them.

L3n41c replied:
Indeed, I noticed that this pattern was common in our code base.
I’ve also noticed that the [`perfsprint` linter](https://github.com/catenacyber/perfsprint), [included in latest versions of `golangci-lint`](https://golangci-lint.run/usage/linters/#perfsprint) [complains about it](https://golangci-lint.run/usage/linters/#perfsprint).
So, we’ll hopefully have to fix all of them next time we update `golangci-lint`.

L3n41c commented on comp/core/tagger/global.go:
```diff
@@ -44,8 +44,25 @@ func GetEntity(entityID types.EntityID) (*types.Entity, error) {
+
+	entityID, err := types.NewEntityIDFromString(entity)
+	if err != nil {
+		return []string{}, err
```
Does it make a difference to return an empty slice instead of `nil` (like what is done above line 53) in case error ?

```suggestion
		return nil, err
```

L3n41c commented on pkg/collector/corechecks/ebpf/oomkill/oom_kill.go:
```diff
@@ -115,9 +115,9 @@ func (m *OOMKillCheck) Run() error {
+		entityID := types.NewEntityID(types.ContainerID, containerID)
 		var tags []string
-		if entityID != "" {
+		if entityID.Empty() {
```
This change looks counter-intuitive to me because I would have expected `""` to be equivalent to `.Empty()` and not the other way round.

Shouldn’t this be:
```suggestion
		if entityID.NotEmpty() {
```
or
```suggestion
		if !entityID.Empty() {
```
?

L3n41c commented on pkg/collector/corechecks/ebpf/tcpqueuelength/tcp_queue_length.go:
```diff
@@ -110,9 +110,9 @@ func (t *TCPQueueLengthCheck) Run() error {
+		entityID := types.NewEntityID(types.ContainerID, containerID)
 		var tags []string
-		if entityID != "" {
+		if entityID.Empty() {
```
This change looks counter-intuitive to me because I would have expected `""` to be equivalent to `.Empty()` and not the other way round.

Shouldn’t this be:
```suggestion
		if entityID.NotEmpty() {
```
or
```suggestion
		if !entityID.Empty() {
```
?

adel121 committed:
address PR comments - mainly remove all calls to NewEntityIDFromString

adel121 commented:
@L3n41c 

thank you for you insightful comments and remarks. 

I addressed all your comments in [this](https://github.com/DataDog/datadog-agent/pull/30133/commits/e6a0deb8b8e632978cbf3829413d5f0d07d7977a) commit. 

Let me know if you find them satisfactory 🙇

jackgopack4 commented:
If I want to "rip the bandaid off" what changes do I need to make to code that you modified to call LegacyTag?

davidor commented on comp/otelcol/otlp/collector.go:
```diff
@@ -70,7 +70,7 @@ func (t *tagEnricher) Enrich(_ context.Context, extraTags []string, dimensions *
 	enrichedTags = append(enrichedTags, dimensions.Tags()...)
 
-	entityTags, err := tagger.Tag(dimensions.OriginID(), t.cardinality)
+	entityTags, err := tagger.LegacyTag(dimensions.OriginID(), t.cardinality)
```
Shouldn't we try to convert to entityID here to avoid calling `LegacyTag` from Go code?

adel121 replied:
Right, updated

adel121 committed:
pr review

jackgopack4 reviewed (approved):
this makes sense; always a fan of separating out dependencies especially since tagger is such a strong global component

adel121 committed:
run inv -e go-deps.generate

louis-cqrl reviewed (approved):
LGTM for file ASC owns

louis-cqrl reviewed (approved):
LGTM for file ASC owns

adel121 committed:
fix lint

adel121 commented:
/merge
